compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp

```diff
@@ -0,0 +1,202 @@
+    switch (currSize) {
+    case 0:
+    case 1:
+    case 2:
+      mapping.push_back(IREE::Codegen::WorkgroupMappingAttr::get(
+          context, IREE::Codegen::symbolizeWorkgroupId(currSize).value()));
+      break;
+    default:
+      mapping.push_back(IREE::Codegen::WorkgroupMappingAttr::get(
+          context, IREE::Codegen::WorkgroupId::IdZ, currSize - 2));
+    }
+  }
+  return llvm::to_vector(llvm::reverse(mapping));
+}
+
+void TileAndDistributeToWorkgroupsUsingForallOpPass::runOnOperation() {
+  auto funcOp = getOperation();
+  auto *context = &getContext();
+  SmallVector<Operation *> computeOps = getComputeOps(funcOp);
+
+  IRRewriter rewriter(context);
+  FailureOr<TilingInfo> tilingInfo =
+      getTiledAndDistributionInfo(rewriter, computeOps);
+  if (failed(tilingInfo)) {
+    return signalPassFailure();
+  }
+  auto tilableOp = dyn_cast_or_null<TilingInterface>(tilingInfo->tilableOp);
+  if (!tilableOp) {
+    // Did not find a tileable op. So do nothing.
+    return;
+  }
+
+  scf::SCFTilingOptions tilingOptions;
+  tilingOptions.setTileSizes(tilingInfo->tileSizes);
+  tilingOptions.setInterchange(tilingInfo->interchange);
+  tilingOptions.setLoopType(scf::SCFTilingOptions::LoopType::ForallOp);
+  SmallVector<Attribute> deviceMappingAttribute =
+      getMapping(context, tilingInfo->tileSizes);
+  if (failed(IREE::Codegen::WorkgroupMappingAttr::verifyAttrList(
+          context, ::mlir::detail::getDefaultDiagnosticEmitFn(funcOp.getLoc()),
+          deviceMappingAttribute))) {
+    return signalPassFailure();
+  }
+  tilingOptions.setMapping(deviceMappingAttribute);
+
+  scf::SCFTileAndFuseOptions tileAndFuseOptions;
+  tileAndFuseOptions.setTilingOptions(tilingOptions);
+  // TODO: For now use the default tile and fuse control function. That needs
+  // to be modified to allow for returning the values of the producer when
+  // needed.
+  rewriter.setInsertionPoint(tilableOp);
+
+  // If the `tilableOp` is a `memref` op, then just tile the operation.
+  SmallVector<LoopLikeOpInterface> tilingLoops;
+  if (tilableOp->getNumResults() == 0) {
+    FailureOr<scf::SCFTilingResult> tilingResult =
+        scf::tileUsingSCF(rewriter, tilableOp, tilingOptions);
+    if (failed(tilingResult)) {
+      funcOp.emitOpError("tiling failed");
+      return signalPassFailure();
+    }
+    rewriter.eraseOp(tilableOp);
+    std::swap(tilingResult->loops, tilingLoops);
+  } else {
+    FailureOr<scf::SCFTileAndFuseResult> tileAndFuseResult =
+        scf::tileConsumerAndFuseProducersUsingSCF(rewriter, tilableOp,
+                                                  tileAndFuseOptions);
+    if (failed(tileAndFuseResult)) {
+      funcOp.emitOpError("tile and fuse greedily failed");
+      return signalPassFailure();
+    }
+    for (auto [origValue, replacement] : tileAndFuseResult->replacements) {
+      rewriter.replaceAllUsesWith(origValue, replacement);
+    }
+    std::swap(tileAndFuseResult->loops, tilingLoops);
+  }
+
```

Comment:
Do you plan to add the consumer fusion to the pass? I think we'll need that to stop lowering_config propagation.

Reply:
Yes, I want to do that as a follow up.

Reply:
@MaheshRavishankar If you want me to add the consumer fusion -- I would be happy to help.

Reply:
Yeah, that would be great. I think if you can combine what you added in the PR for tile and fuse pass you added, it would be great.